ember-core/src/main/scala/org/http4s/ember/core/Parser.scala

```diff
@@ -0,0 +1,270 @@
+      for {
+        line <- bv.decodeAscii.toOption
+        // _ = println(s"Generate Headers - line: ${line}")
+        idx <- Some(line.indexOf(':'))
+        if idx >= 0
+        header = Header(line.substring(0, idx), line.substring(idx + 1).trim)
+        // _ = println(s"generateHeaders - header: ${header}")
+      } yield header
+
+    headerO match {
+      case Right((lineBV, rest)) =>
+        val headerO = generateHeaderForLine(lineBV)
+        val newHeaders = acc ++ headerO.map(Headers.of(_)).foldMap(identity)
+        // println(s"Generate Headers Header0 = $headerO")
+        generateHeaders(rest)(newHeaders)
+      case Left(bv) =>
+        val headerO = generateHeaderForLine(bv)
+        headerO.map(Headers.of(_)).foldMap(identity) ++ acc
+    }
+
+  }
+
+  def splitHeader(byteVector: ByteVector): Either[ByteVector, (ByteVector, ByteVector)] = {
+    val index = byteVector.indexOfSlice(`\r\n`)
+    if (index >= 0L) {
+      val (line, rest) = byteVector.splitAt(index)
+      // println(s"splitHeader - Slice: ${line.decodeAscii}- Rest: ${rest.decodeAscii}")
+      Either.right((line, rest.drop(`\r\n`.length)))
+    } else {
+      Either.left(byteVector)
+    }
+  }
+
+  object Request {
+
+    def parser[F[_]: Sync](maxHeaderLength: Int)(s: Stream[F, Byte]): F[Request[F]] =
+      s.through(httpHeaderAndBody[F](maxHeaderLength))
+        .evalMap { case (bv, body) => headerBlobByteVectorToRequest[F](bv, body, maxHeaderLength) }
+        .take(1)
+        .compile
+        .lastOrError
+
+    private def headerBlobByteVectorToRequest[F[_]: MonadError[?[_], Throwable]](
+        b: ByteVector,
+        s: Stream[F, Byte],
+        maxHeaderLength: Int): F[Request[F]] =
+      for {
+
+        (methodHttpUri, headersBV) <- splitHeader(b).fold(
+          _ =>
+            ApplicativeError[F, Throwable].raiseError[(ByteVector, ByteVector)](
+              EmberException.ParseError("Invalid Empty Init Line")),
+          Applicative[F].pure(_)
+        )
+        // Raw Header Logging
+        // _ = println(s"HeadersSection - ${headersBV.decodeAscii}")
+
+        headers = generateHeaders(headersBV)(Headers.empty)
+        // _ = println(headers)
+
+        (method, uri, http) <- bvToRequestTopLine[F](methodHttpUri)
+
+        contentLength = headers.get(org.http4s.headers.`Content-Length`).map(_.length).getOrElse(0L)
+        host = headers.get(org.http4s.headers.Host)
+        isChunked = headers
+          .get(org.http4s.headers.`Transfer-Encoding`)
+          .exists(_.value.toList.contains(TransferCoding.chunked))
+
+        body = Alternative[Option]
+          .guard(isChunked)
+          .fold(
+            s.take(contentLength)
+          )(_ => s.through(ChunkedEncoding.decode(maxHeaderLength)))
+
+        // enriched with host
+        // seems presumptious
+        newUri = uri.copy(
+          authority = host.map(h => Uri.Authority(host = Uri.RegName(h.host), port = h.port)))
```

Comment:
That `RegName` might be an IP address, no?  I think this is something the URI work needs to make easier to get right.

Reply:
Yeah, this is  place that feel uncomfortable.

Reply:
I think we have a few latent bugs related to this.  I'll take it on myself to come back and clean this one up.